Created a very simple Set struct that only accepts Int type

andresnowak opened the pull request.

andresnowak committed:
Created a very simple SetInt struct

Moosems commented on types/set/test_set.mojo:
Should we make a `test` folder next to `types`?

andresnowak replied:
if you want i can put it there

andresnowak replied:
I just don't know how do you import packages that are outside its folder

andresnowak replied:
It seems it is not possible for now "*Currently, top-level code is not supported in .mojo files, so unlike Python, you can’t write code in __init__.mojo*"

Moosems commented on types/set/README.md:
```diff
@@ -0,0 +1,137 @@
+
+> `fn __init__(inout self) -> None:`
+
+Constructs a new `SetInt` with a default capacity (capacity_default = 11).
```
May I ask why the default is 11 rather than a nice round number like 10? Not annoyed or anything, just curious :D.

andresnowak replied:
The idea was just to use a prime number to start to reduce a little bit the collisions that can happen (i don't think it works in reality i just used as an idea, because this is a very very simple set)

Moosems commented on types/set/README.md:
```diff
@@ -0,0 +1,137 @@
+
+-   `None`
+
+#### `__init__`
```
Two init sections with the same info? Could we remove the second one?

Moosems commented on types/set/README.md:
```diff
@@ -0,0 +1,137 @@
+
+_Module_
+
+Implements a very simple `SetInt` struct.
```
Could we add a little section explaining its use cases and purpose? I'll do this in the other types too.

Moosems commented on types/set/README.md:
```diff
@@ -0,0 +1,137 @@
+
+-   `None`
+
+#### `remove`
```
What happens if I have two of the same Int in the set and try to remove it? Will it remove all instances?

Moosems replied:
The code shows it deletes all instances (I think) but just adding that to the readme would be nice.

andresnowak replied:
there can only exist one instance of a value in a set, so it only removes one

Moosems commented on types/set/README.md:
````diff
@@ -0,0 +1,137 @@
+
+```mojo
+from types.array import Array
+from type.dodgy import DodgyString
````
Can we remove the unnecessary imports (I don't see it used anywhere but I may be missing something)?

Moosems commented on types/set/set.mojo:
```diff
@@ -0,0 +1,102 @@
+    var data: Pointer[Int]
+    var flag_for_zero_value: Bool
+
+    fn __init__(inout self):
```
Could we make an init where one can give a different size?

Moosems commented on types/set/set.mojo:
```diff
@@ -0,0 +1,102 @@
+from memory.unsafe import DTypePointer
```
I don't see this used anywhere...

Moosems commented on types/set/README.md:
```diff
@@ -0,0 +1,137 @@
+from type.dodgy import DodgyString
+
+fn main():
+    from set import SetInt
```
Could we move this to the top of the example?

Moosems commented on types/set/set.mojo:
```diff
@@ -0,0 +1,102 @@
+        self.size = new_size
+
+    fn print_all(self):
+        print("[")
```
```diff
-        print("[")
+        print_no_newline("[")
```

Moosems commented on types/set/set.mojo:
```diff
@@ -0,0 +1,102 @@
+    fn __resize(inout self):
+        let capacity_ratio = self._filled / self.size
+
+        if capacity_ratio < self.resize_factor_down and self.size // 2 >= self.min_size:
```
Could a few comments be added explaining this?

Moosems commented on types/set/set.mojo:
```diff
@@ -0,0 +1,102 @@
+
+    fn add(inout self, value: Int):
+        if value == 0:
+            self.flag_for_zero_value = True
```
What is wrong with adding an Int with the value 0?

andresnowak replied:
because i am using the zero values as a check to know if there already exists a value in the set or not, so because there are always zero values in the set i use a flag to know if a zero was added.

andresnowak committed:
Fixed the readme

andresnowak committed:
Added comments and removed unused imports

andresnowak committed:
Added purpose in README

Moosems committed:
Rename, remove, update test, small changes

Renamed files, updated main README, removed one time variables, updates test, changed print, approved for push.

Moosems reviewed (approved):
Looks good to me!

andresnowak commented:
okay, then you can merge the branch if you want (i don't have write access, so i can't do it)

Moosems commented:
Thank you for the addition @andresnowak!
